### PR TITLE
Teams enhancements

### PR DESCRIPTION
**Pull request by merit-maita**
## Contributing to Cortex XSOAR Content
Make sure to register your contribution by filling the [contribution registration form](https://forms.gle/XDfxU4E61ZwEESSMA)

**The Pull Request will be reviewed only after the contribution registration form is filled.**

## Status
- [x] In Progress
- [ ] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://jira-hq.paloaltonetworks.local/browse/CIAC-4575

## Description
- Improved implementation of the message sent to chat in Microsoft Teams through data collection task.
- Improved implementation of the ***microsoft-teams-chat-create*** and ***microsoft-teams-message-send-to-chat*** commands to add the bot automatically to the chat.
- Improved implementation of the response returned to the user in an one-to-one chat with the bot.
- Documentation and metadata improvements.
- 
## Screenshots
Paste here any images that will help the reviewer

## Minimum version of Cortex XSOAR
- [ ] 6.0.0
- [ ] 6.1.0
- [ ] 6.2.0
- [ ] 6.5.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [ ] No

## Must have
- [ ] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 1]
```diff
@@ -252,7 +254,10 @@ def urlify_hyperlinks(message: str) -> str:
     # URLify markdown hyperlinks
     urls = re.findall(URL_REGEX, message)
     for url in urls:
-        formatted_message = formatted_message.replace(url, f'[{url}]({url})')
+        if EXTERNAL_FORM in url:
+            formatted_message = formatted_message.replace(url, f'[Microsoft Teams Form]({url})')
```
Comment (DeanArbel) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:258]: Since this is a special use case, please add a short docstring to explain where the external_form is coming from.

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 2]
```diff
@@ -1274,6 +1279,38 @@ def channel_user_list_command():
     return_results(result)
 
 
+def is_bot_in_chat(chat_id):
+    """
+    check if the bot is already in the chat.
+    """
+    url_suffix = f"v1.0/chats/{chat_id}/installedApps?$expand=teamsApp," \
+                 f"teamsAppDefinition&$filter=teamsApp/externalId eq '{BOT_ID}'"
```
Comment (DeanArbel) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:1287]: Please use params instead of explicitly writing the params as part of the suffix.

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 3]
```diff
@@ -1274,6 +1279,38 @@ def channel_user_list_command():
...
+    :param chat_id: chat id which to add the bot to.
+    """
+
+    demisto.debug(f'adding bot with id {BOT_ID} to chat')
+
+    # bot is already part of the chat
+    if is_bot_in_chat(chat_id):
+        return
+
+    res = http_request('GET', f"{GRAPH_BASE_URL}/v1.0/appCatalogs/teamsApps?$filter=externalId eq '{BOT_ID}'")
```
Comment (DeanArbel) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:1304]: Please use params instead of explicitly writing the params as part of the suffix.

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 4]
```diff
@@ -2141,35 +2206,38 @@ def direct_message_handler(integration_context: dict, request_body: dict, conver
...
-            data = 'You are not allowed to create incidents.'
-        else:
+    lowered_message = message.lower()
+    # the command is to create new incident
+    create_incident = lowered_message.find('incident') != -1 and (lowered_message.find('create') != -1
+                                                                  or lowered_message.find('open') != -1
+                                                                  or lowered_message.find('new') != -1)
+    data = ""
+    if not demisto_user:
+        data = handle_external_user(user_email if user_email else username,
```
Comment (DeanArbel) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:2218]: you can use to `or` operator to simplify this.
```suggestion
        data = handle_external_user(user_email or username,
```

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 5]
```diff
@@ -2247,9 +2315,10 @@ def message_handler(integration_context: dict, request_body: dict, channel_data:
                             investigation_id: str = mirrored_channel.get('investigation_id', '')
                             username: str = from_property.get('name', '')
                             user_email: str = get_team_member(integration_context, team_member_id).get('user_email', '')
+                            demisto.debug(f"Adding Entry {message} to investigation {investigation_id}")
                             demisto.addEntry(
                                 id=investigation_id,
-                                entry=message,
+                                entry=message.strip(),
```
Comment (DeanArbel) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:2321]: The original intent was to use `lstrip` to handle unexpected spaces introduced, but `strip` also makes sense.
Please add a short docstring to explain this.
This should be tested via a unit test.

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 6]
```diff
@@ -2263,13 +2332,15 @@ def messages() -> Response:
...
     try:
+        demisto.debug("Microsoft Teams Integration received a message from Teams")
         demisto.debug('Processing POST query...')
         headers: dict = cast(Dict[Any, Any], request.headers)
 
         if validate_auth_header(headers) is False:
             demisto.info(f'Authorization header failed: {str(headers)}')
         else:
             request_body: dict = request.json
+            demisto.debug(f"request body: {request_body}")
```
Comment (DeanArbel) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:2343]: Can this contain any sensitive data?

**Review by DeanArbel — CHANGES_REQUESTED** [review]

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py (hunk 2 above)
Comment (merit-maita) [reply Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:1287]: i don't understand exactly what you mean

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py (hunk 3 above)
Comment (merit-maita) [reply Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:1304]: same as above

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/README.md [hunk 7]
```diff
@@ -1,7 +1,10 @@
 Use the Microsoft Teams integration to send messages and notifications to your team members and create meetings.
 This integration was integrated and tested with version 1.0 of Microsoft Teams.
 
-**Note:** For use cases where it is only need to send messages to a specific channel, we recommend checking the [Microsoft Teams via Webhook Integration](https://xsoar.pan.dev/docs/reference/integrations/microsoft-teams-via-webhook), which has a simpler setup.
+**Note::** 
+- The integration has the ability to run builtin XSOAR commands, through a mirrored channel. make sure to pass the command in the chat exactly as typed in CORTEX XSOAR CLI. example: `!DeleteContext all=yes`. Use command `mirror-investigation` to mirror/create a mirrored channel.
```
Comment (ShirleyDenkberg) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/README.md:5]:
```suggestion
- The integration has the ability to run built-in Cortex XSOAR commands, through a mirrored channel. Make sure to pass the command in the chat exactly as typed in the CORTEX XSOAR CLI. For example: `!DeleteContext all=yes`. Use the command `mirror-investigation` to mirror/create a mirrored channel.
```

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/README.md [hunk 8]
```diff
@@ -1,7 +1,10 @@
...
+
+- For use cases where it is only need to send messages to a specific channel, we recommend checking the [Microsoft Teams via Webhook Integration](https://xsoar.pan.dev/docs/reference/integrations/microsoft-teams-via-webhook), which has a simpler setup.
```
Comment (ShirleyDenkberg) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/README.md:7]:
```suggestion
- For use cases where it is only needed to send messages to a specific channel, we recommend checking the [Microsoft Teams via Webhook Integration](https://xsoar.pan.dev/docs/reference/integrations/microsoft-teams-via-webhook), which has a simpler setup.
```

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/README.md [hunk 9]
```diff
@@ -348,6 +351,7 @@ https://login.microsoftonline.com/TENANT_ID/oauth2/v2.0/authorize?response_type=
 - Posting a message or adaptive card to a private/shared channel is currently not supported in the ***send-notification*** command. Thus, also the ***mirror_investigation*** command does not support private/shared channels. For more information, see [Microsoft General known issues and limitations](https://learn.microsoft.com/en-us/connectors/teams/#general-known-issues-and-limitations).
 - In case of multiple chats/users sharing the same name, the first one will be taken.
 - See Microsoft documentation for [Limits and specifications for Microsoft Teams](https://learn.microsoft.com/en-us/microsoftteams/limits-specifications-teams).
+- If a non XSOAR user ran the command `new incident` in the chat with the bot, the owner of the created incident would be the logged in XSOAR user, not the external user who ran the command.
```
Comment (ShirleyDenkberg) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/README.md:354]:
```suggestion
- If a non-Cortex XSOAR user ran the `new incident` command in the chat with the bot, the owner of the created incident would be the logged in Cortex XSOAR user, not the external user who ran the command.
```

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/README.md [hunk 10]
```diff
@@ -393,6 +397,8 @@ Message was sent successfully.
 ***
 Mirrors the Cortex XSOAR investigation to the specified Microsoft Teams channel. Supports only standard channels.
 
+**Note**: Mirrored channels could be used to run CORTEX XSOAR builtin commands.
```
Comment (ShirleyDenkberg) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/README.md:400]:
```suggestion
**Note**: Mirrored channels could be used to run Cortex XSOAR built-in commands.
```

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/README.md [hunk 11]
```diff
@@ -1130,6 +1144,8 @@ You can send the message `help` in order to see the supported commands:
     
     Try inserting your configured message endpoint in a browser tap, click `Enter`, if `Method Not Allowed` is returned, the endpoint is valid and ready to communicate, otherwise, it needs to be handled according to the returned error's message.
 
+    In some cases, a connection is not created between Teams and the messaging endpoint, when adding a bot to the team. You can work around this problem by adding any member to the team the bot was added to (the bot should be already added to the team). It will trigger a connection and solve the issue, then you can remove the member that was added with no poroblems.
```
Comment (ShirleyDenkberg) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/README.md:1147]:
```suggestion
    In some cases, a connection is not created between Teams and the messaging endpoint, when adding a bot to the team. You can work around this problem by adding any member to the team the bot was added to (the bot should be already added to the team). This will trigger a connection and solve the issue. You can then remove the member that was added.
```

File: Packs/MicrosoftTeams/ReleaseNotes/1_4_10.md [hunk 12]
```diff
@@ -0,0 +1,7 @@
+
+#### Integrations
+##### Microsoft Teams
+- Improved implementation of the message sent to chat in Microsoft Teams through data collection task.
```
Comment (ShirleyDenkberg) [inline Packs/MicrosoftTeams/ReleaseNotes/1_4_10.md:4]:
```suggestion
- Improved implementation of the message sent to chat in Microsoft Teams through the data collection task.
```

File: Packs/MicrosoftTeams/ReleaseNotes/1_4_10.md [hunk 13]
```diff
@@ -0,0 +1,7 @@
...
+- Improved implementation of the ***microsoft-teams-chat-create*** and ***microsoft-teams-message-send-to-chat*** commands to add the bot automatically to the chat.
```
Comment (ShirleyDenkberg) [inline Packs/MicrosoftTeams/ReleaseNotes/1_4_10.md:5]:
```suggestion
- Improved implementation of the following commands to add the bot automatically to the chat.
   - ***microsoft-teams-chat-create***
   - ***microsoft-teams-message-send-to-chat*** 
```

File: Packs/MicrosoftTeams/ReleaseNotes/1_4_10.md [hunk 14]
```diff
@@ -0,0 +1,7 @@
...
+- Improved implementation of the response returned to the user in an one-to-one chat with the bot.
```
Comment (ShirleyDenkberg) [inline Packs/MicrosoftTeams/ReleaseNotes/1_4_10.md:6]:
```suggestion
- Improved implementation of the response returned to the user in a one-to-one chat with the bot.
```

**Comment by ShirleyDenkberg** [comment]
@DeanArbel Doc review completed.

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 15]
```diff
@@ -2114,6 +2159,29 @@ def member_added_handler(integration_context: dict, request_body: dict, channel_
     set_integration_context(integration_context)
 
 
+def handle_external_user(user_identifier, allow_create_incident, create_incident):
```
Comment (bziser) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:2162]: typing :)

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 16]
```diff
@@ -1274,6 +1280,40 @@ def channel_user_list_command():
...
+    """
+
+    url_suffix = f"v1.0/chats/{chat_id}/installedApps"
+    res = http_request('GET', urljoin(GRAPH_BASE_URL, url_suffix),
+                       params={"$expand": "teamsApp,teamsAppDefinition",
+                               "$filter": "teamsApp/externalId eq '{BOT_ID}'"})
+    return True if res.get('value') else False      # type: ignore
+
+
+def add_bot_to_chat(chat_id: str):  # pragma: no cover
```
Comment (bziser) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:1295]: return type :)

**Review by bziser — COMMENTED** [review]
👏

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py (hunk 16 above)
Comment (merit-maita) [reply Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:1295]: it does not return anything

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 17]
```diff
@@ -683,7 +689,7 @@ def integration_health():
     })
 
 
-def validate_auth_header(headers: dict) -> bool:
+def validate_auth_header(headers: dict) -> bool:    # pragma: no cover
```
Comment (bziser) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:692]: Why did you add `# pragma: no cover` on this function?

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 18]
```diff
@@ -170,7 +172,7 @@ def create_incidents(demisto_user: dict, incidents: list) -> dict:
     return data
 
 
-def process_incident_create_message(demisto_user: dict, message: str) -> str:
+def process_incident_create_message(demisto_user: dict, message: str) -> str:   # pragma: no cover
```
Comment (bziser) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:175]: Why did you add `# pragma: no cover` on this function?

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py (hunk 17 above)
Comment (merit-maita) [reply Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:692]: it was all red in the coverage report

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py (hunk 18 above)
Comment (merit-maita) [reply Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:175]: same as above

File: Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py [hunk 19]
```diff
@@ -1303,6 +1343,8 @@ def chat_create_command():
     chat_data.pop('@odata.context', '')
     chat_data['chatId'] = chat_data.pop('id', '')
 
+    add_bot_to_chat(chat_data.get("chatId"))    # type: ignore
```
Comment (bziser) [inline Packs/MicrosoftTeams/Integrations/MicrosoftTeams/MicrosoftTeams.py:1346]:
```suggestion
    add_bot_to_chat(chat_data.get("chatId", ''))
```

**Review by bziser — APPROVED** [review]
Missed that one :)
